java-checks/src/main/java/org/sonar/java/checks/ReturnInFinallyCheck.java

```diff
@@ -52,53 +58,131 @@ public void visitTryStatement(TryStatementTree tree) {
-    isInFinally.push(false);
+    isInFinally.push(tree.kind());
     super.visitMethod(tree);
     isInFinally.pop();
   }
 
   @Override
+  public void visitForStatement(ForStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitForStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitForEachStatement(ForEachStatement tree) {
+    isInFinally.push(tree.kind());
+    super.visitForEachStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitWhileStatement(WhileStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitWhileStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitDoWhileStatement(DoWhileStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitDoWhileStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
+  public void visitSwitchStatement(SwitchStatementTree tree) {
+    isInFinally.push(tree.kind());
+    super.visitSwitchStatement(tree);
+    isInFinally.pop();
+  }
+
+  @Override
   public void visitReturnStatement(ReturnStatementTree tree) {
-    reportIssue(tree.returnKeyword(), "return");
+    reportIssue(tree.returnKeyword(), tree.kind());
     super.visitReturnStatement(tree);
   }
 
   @Override
   public void visitThrowStatement(ThrowStatementTree tree) {
-    reportIssue(tree.throwKeyword(), "throw");
+    reportIssue(tree.throwKeyword(), tree.kind());
     super.visitThrowStatement(tree);
   }
 
   @Override
   public void visitContinueStatement(ContinueStatementTree tree) {
-    reportIssue(tree.continueKeyword(), "continue");
+    reportIssue(tree.continueKeyword(), tree.kind());
     super.visitContinueStatement(tree);
   }
 
   @Override
   public void visitBreakStatement(BreakStatementTree tree) {
-    reportIssue(tree.breakKeyword(), "break");
+    reportIssue(tree.breakKeyword(), tree.kind());
     super.visitBreakStatement(tree);
   }
 
-  private void reportIssue(Tree tree, String statement) {
-    if (isInFinally()) {
-      context.reportIssue(this, tree, "Remove this "+statement+" statement from this finally block.");
+  private void reportIssue(Tree tree, Tree.Kind jumpKind) {
+    if (isAbruptFinallyBlock(jumpKind)) {
+      context.reportIssue(this, tree, "Remove this " + statement(jumpKind) + " statement from this finally block.");
     }
   }
 
-  private boolean isInFinally() {
-    return !isInFinally.isEmpty() && isInFinally.peek();
+  private static String statement(Tree.Kind kind) {
+    switch (kind) {
+      case BREAK_STATEMENT:
+        return "break";
+      case CONTINUE_STATEMENT:
+        return "continue";
+      case RETURN_STATEMENT:
+        return "return";
+      case THROW_STATEMENT:
+        return "throw";
+      default:
+        throw new IllegalArgumentException("Unsupported jump statement!");
+    }
   }
 
+  private boolean isAbruptFinallyBlock(Tree.Kind jumpKind) {
+    if (isInFinally.isEmpty()) {
+      return false;
```

Comment:
It was covered even before here https://github.com/SonarSource/sonar-java/blob/97306da9245167ec3f6038b1ed0dc3d193990e5a/java-checks/src/test/files/checks/ReturnInFinallyCheck.java#L64-L64  . It is bit hidden though.